[TAN-943] Add layer_type and url to map_layers

jinjagit opened the pull request:
Uses STI to introduce 2 new child classes to the `Layer` model: `GeojsonLayer` (representing the layer types we currently use) and `EsriFeatureLayer` for the new layer type we are introducing. Only the 2 new model classes are now used to create records.

- Adds new attributes to `maps_layers` table & introduces 2 new models, as children of `CustomMaps::Layer`
- Adds basic validations for these new attributes
- Adds basic serialization and related permitted params.
- Adds model & acceptance tests (and refactors the acceptance tests)
- Adds the new attributes to `ProjectCopyService#yml_maps_layers`
- Adds the new attributes to `MultiTenancy::Templates::Serializers:CustomMaps::Layer`
- Adds db level default of empty hash to `geojson` attribute.

## Implementation
Esri Feature layers, and possibly other layer types (yet to be defined/implemented), will need a URL that points to the layer data on an external API. This URL would be provided by the customer, via the BO create/edit map UI.

# Changelog
## Technical
- [TAN-893] Add 2 new types of map layers: GeojsonLayer (current implementation) & EsriFeatureLayer (new feature, in development)

jinjagit committed:
[TAN-943] Add migration to engine and copy into main db/migrate

jinjagit committed:
[TAN-943] Rename type to layer_type. type is reserved name

jinjagit committed:
[TAN-943] migrate

jinjagit committed:
[TAN-943] basic serialization and permit params

jinjagit committed:
[TAN-943] add new attributes to layer factory

jinjagit committed:
[TAN-943] basic validation of layer_type

jinjagit committed:
[TAN-943] Add new fields to ProjectCopyService yml_maps_layers

jinjagit committed:
[TAN-943] Add basic tests of layer_type validation

jinjagit committed:
[TAN-943] Validate url + basic tests

jinjagit commented on back/engines/commercial/custom_maps/app/models/custom_maps/layer.rb:
```diff
@@ -40,6 +44,10 @@ class Layer < ApplicationRecord
+    validates :layer_type, presence: true, inclusion: { in: LAYER_TYPES }
+    validates :url,
+      format: { with: %r{\Ahttp(s)?://.+\z}, message: 'should start with http:// or https://' },
+      allow_nil: true
```
Unsure if we will need to permit `http://`, but will later limit to only `https://` if we find `http://` is not needed.

Since the URL will be passed to an external API, and should refer to a map layer already uploaded on the external service, we should not be concerned with malicious URLs (e.g. spam) and hence, whitelisting is not needed.

jinjagit committed:
[TAN-943] Validate presence of either url or geojson data, depending on layer_type

jinjagit committed:
[TAN-943] Use explicit regex matching to validate URL

jinjagit committed:
[TAN-943] Test of empty JSON object

jinjagit committed:
[TAN-943] Reset debug version of tests

jinjagit committed:
[TAN-943] Add new attributes to acceptance tests of GET

jinjagit committed:
[TAN-943] Permit not null geojson column and remigrate

jinjagit committed:
[TAN-943] Add acceptance tests to cover new attributes

jinjagit committed:
[TAN-943] Add new attributes to template layer serializer

sebastienhoorens commented on back/db/migrate/20240126095830_add_type_and_url_to_maps_layers.custom_maps.rb:
```diff
@@ -0,0 +1,10 @@
+class AddTypeAndUrlToMapsLayers < ActiveRecord::Migration[7.0]
+  def change
+    add_column :maps_layers, :layer_type, :string, null: false, default: 'geojson'
+    add_column :maps_layers, :url, :string
```
I wonder if `url` is not too vague (as there is also a `marker_svg_url `), perhaps something like `esri_url` could be more specific?

jinjagit replied:
I agree. I had the same thought... would be easier to find in code too.

sebastienhoorens commented on back/engines/commercial/custom_maps/spec/acceptance/map_layers_spec.rb:
```diff
@@ -110,70 +114,107 @@
     patch 'web_api/v1/projects/:project_id/map_config/layers/:id' do
       with_options scope: :layer, required: true, with_example: true do
         parameter :title_multiloc,  'The name of the layer in multiple locales'
+        parameter :layer_type,      'The type of the layer (geojson or esri_feature_service)', required: true
```
Do we want to support changing the layer type of an existing layer? Will this ever happen?

jinjagit replied:
Very good point 👍

sebastienhoorens reviewed (approved):
All good. I have a few comments, but nothing blocking at all.

I wonder, what do the ESRI URLs return? If they return structures that describe the layers, is there a reason that we don't store those in the DB instead of the URL (perhaps it stay up to date that way)?

I have mixed feelings about removing the NOT NULL constraint. It makes sense, but perhaps we could also use the empty hash as default. Metabase doesn't work well when null values can occur for json attributes (but this shouldn't affect our modeling decisions).

Is there a reason to not use STI (single table inheritance) for the layer types? An advantage is that this could avoid conditionals on the layer type. I'm fine with either approach.

jinjagit commented:
> Is there a reason to not use STI (single table inheritance) for the layer types? An advantage is that this could avoid conditionals on the layer type. I'm fine with either approach.

I thought you'd say this! I think, on balance, STI indeed makes more sense, especially as there is every chance we add more layer types as mapping gets more sophisticated. I was undecided, but was suspecting you would help me decide.

I'll switch to STI. Thanks!

I also think your solution of an empty hash for the `geojson` default could work well.

Thanks for the very useful review.

jinjagit committed:
[TAN-943] Rename Layer.url to Layer.layer_url

jinjagit committed:
[TAN-943] Replace geojson not null constraint + add default of empty json hash

jinjagit committed:
[TAN943] Add comment

jinjagit committed:
[TAN-943] Use STI pattern

jinjagit committed:
[TAN-943] Don't permit type param on update action, and add tests

jinjagit committed:
[TAN-943] Avoid setting default type + set type for existing records in migration

jinjagit committed:
[TAN-943] Replace with_layers trait with with_geojson_layers trait

jinjagit committed:
[TAN-222] Fix tenant_serializer_spec by including Layer type in seeds

jinjagit committed:
[TAN-222] Fix apply_service_spec by including Layer type in e2etests_template

jinjagit committed:
TAN-943] Improve test

jinjagit commented:
@sebastienhoorens 

- Now using STI (=> no conditional validations)
- default empty hash for `geojson` attribute
- `type` param not permitted for update action (cannot change type of record via endpoint)
- Changed from setting a default layer `type` in migration, to no default and updating all existing layers to be `type: 'GeojsonLayer'` (I think this is less likely to produce unexpected behavior)

jinjagit committed:
[TAN-943] Improve comment

sebastienhoorens commented on back/engines/commercial/custom_maps/app/controllers/custom_maps/web_api/v1/layers_controller.rb:
```diff
@@ -52,21 +52,34 @@ def reorder
+
         def serialized_layer
-          CustomMaps::WebApi::V1::LayerSerializer.new(@layer).serializable_hash.to_json
+          "CustomMaps::WebApi::V1::#{@layer.type.demodulize}Serializer".constantize.new(@layer).serializable_hash.to_json
```
This is considered a little bit dangerous. It's very hypothetical, but suppose there would exist a `DeleteAllDataOnInstantiationSerializer` then this class could get instantiated by passing a corresponding value for the type in the API call. In practice there is probably no danger at all. But it could be worth considering to instead write something like:
```
serializer_class = case @layer.type
  GeoJsonSerializer when 'GeoJSON'
  EsriSerializer when 'ESRI'
  end
```

jinjagit replied:
Good idea. I was aware of the issue of malicious class calling via the API, but thought the whitelisting of the `type` in the `Layer` model was sufficient to prevent it.

I'll add your explicit conditional approach into the method, however, for extra security 👍

sebastienhoorens reviewed (approved):
All good, just one small remark.

jinjagit committed:
[TAN-943] Use explicit matching of param string to set layer serializer

This is more secure than interpolating a user provided string to create
a class name to be instantiated.

jinjagit committed:
[TAN-943] Rename layer to EsriFeatureLayer

jinjagit committed:
Merge branch 'master' into TAN-943-add-type-and-url-to-layers

jinjagit committed:
[TAN-943] Typo in comment

jinjagit committed:
[TAN-943] Remove unrelated FE change (unsure how this crept into the PR)

jinjagit commented on front/app/api/analysis_insights/useAnalysisInsights.ts:
```diff
@@ -7,7 +7,7 @@ import { IInsights, InsightsKeys, IInsightsParams } from './types';
   return fetcher<IInsights>({
     path: `/analyses/${analysisId}/insights`,
-    action: 'get'
+    action: 'get',
```
I'm assuming some kind of auto-linter is adding this in, as I didn't and it gets added back when I remove it!

jinjagit committed:
[TAN-943] Improve copy in AdminHQ settings

jinjagit committed:
Merge branch 'master' into TAN-943-add-type-and-url-to-layers

amanda-anderson committed:
Add new type to FE

amanda-anderson committed:
Update tests

jinjagit committed:
Merge branch 'master' into TAN-943-add-type-and-url-to-layers

jinjagit committed:
[TAN-943] Replace legacy admin_api serializer

jinjagit committed:
[TAN-943] Move legacy admin_api serializer

jinjagit committed:
[TAN-943] Set map center coords and zoom in create_tenant.rake

jinjagit committed:
[TAN-943] Remove legacy serializer (again)

jinjagit committed:
Merge branch 'master' into TAN-943-add-type-and-url-to-layers

jinjagit committed:
Merge branch 'master' into TAN-943-add-type-and-url-to-layers

jinjagit committed:
Merge branch 'master' into TAN-943-add-type-and-url-to-layers

jinjagit committed:
[TAN-943] manual correction of structure.sql

Added back migration listing of one removed in faulty manual
erge conflict resolution.
